source/features/clone-branch.tsx

```diff
@@ -0,0 +1,89 @@
+import delegate, {DelegateEvent} from 'delegate-it';
+import gitBranch from 'octicon/git-branch.svg';
+import insertTextTextarea from 'insert-text-textarea';
+import React from 'dom-chef';
+import select from 'select-dom';
+import * as api from '../libs/api';
+import features from '../libs/features';
+import {getRepoURL} from '../libs/utils';
+
+async function cloneBranch(event: DelegateEvent<MouseEvent, HTMLAnchorElement>): Promise<void> {
+	const currentTarget = event.delegateTarget;
+	const branchElement = currentTarget.closest<HTMLAnchorElement>('[data-branch-name]');
+	const spinner = select('.js-loading-spinner', branchElement!)!;
+	spinner.hidden = false;
+	currentTarget.hidden = true;
+
+	let newBranchName = prompt('Enter the new branch name')?.trim();
+	if (!newBranchName) {
+		return;
+	}
+
+	const getBranchInfo = await api.v3(`repos/${getRepoURL()}/git/refs/heads/${branchElement!.dataset.branchName!}`);
+	let createBranch = await createNewBranch(newBranchName, getBranchInfo);
+
+	while (Number(createBranch.httpStatus) === 422) {
+		newBranchName = prompt(String(createBranch.message) + '\n Enter the new branch name', newBranchName)!;
+		if (!newBranchName) {
+			break;
+		}
+
+		createBranch = await createNewBranch(newBranchName, getBranchInfo); // eslint-disable-line no-await-in-loop
+	}
+
+	spinner.hidden = true;
+	currentTarget.hidden = false;
+
+	if (!newBranchName) {
+		return;
+	}
+
+	const searchField = select<HTMLInputElement>('.js-branch-search-field')!;
+	searchField.select();
+	insertTextTextarea(searchField, newBranchName);
+}
+
+async function createNewBranch(newBranchName: string, getBranchInfo: AnyObject): Promise<AnyObject> {
+	const createBranch = await api.v3(`repos/${getRepoURL()}/git/refs`, {
+		method: 'POST',
+		body: {
+			sha: getBranchInfo.object.sha,
+			ref: 'refs/heads/' + newBranchName
+		},
+		ignoreHTTPStatus: true
+	});
+	return createBranch;
```

Comment:
1. I think we can place the spinner in this function, because this is the action the user will be waiting for.
2. Only return `true | string`. True is success, string is the error. _Any_ error counts, not just 422

Reply:
I lost you on this one

Reply:
1. Show the spinner while before the API call, hide it after the call, inside this function
2. Parse the result of `api.*` calls before returning them, I told you before. Here it will be something like 

``` js
if response.ok {
  return true
}

return response.message
```

Reply:
If I want to put the spinner there then I need to pass both the current element and the target to the function is that what you want?

Reply:
I think Im missing something. I di what you asked but... if the cancel (They dont put in a new branch) it will just keep running.